src/ecmult_const_impl.h

```diff
@@ -228,4 +228,127 @@ static void secp256k1_ecmult_const(secp256k1_gej *r, const secp256k1_ge *a, cons
+     * on a Weierstrass curve:
+     * https://mailarchive.ietf.org/arch/msg/cfrg/7DyYY6gg32wDgHAhgSb6XxMDlJA/
+     *
+     *
+     * === Background: the effective affine technique ===
+     *
+     * Let phi_u be the isomorphism that maps (x, y) on secp256k1 curve y^2 = x^3 + 7 to
+     * x' = u^2*x, y' = u^3*y on curve y'^2 = x'^3 + u^6*7. This new curve has the same order as
+     * the original (it is isomorphic), but moreover, has the same addition/doubling formulas, as
+     * the curve b=7 coefficient does not appear in those formulas (or at least does not appear in
+     * the formulas implemented in this codebase, both affine and Jacobian). See also Example 9.5.2
+     * in https://www.math.auckland.ac.nz/~sgal018/crypto-book/ch9.pdf.
+     *
+     * This means any linear combination of secp256k1 points can be computed by applying phi_u
+     * (with non-zero u) on all input points (including the generator, if used), computing the
+     * linear combination on the isomorphic curve (using the same group laws), and then applying
+     * phi_u^{-1} to get back to secp256k1.
+     *
+     * Switching to Jacobian coordinates, note that phi_u applied to (X, Y, Z) is simply
+     * (X, Y, Z/u). Thus, if we want to compute (X1, Y1, Z) + (X2, Y2, Z), with identical Z
+     * coordinates, we can use phi_Z to transform it to (X1, Y1, 1) + (X2, Y2, 1) on an isomorphic
+     * curve where the affine addition formula can be used instead.
+     * If (X3, Y3, Z3) = (X1, Y1) + (X2, Y2) on that curve, then our answer on secp256k1 is
+     * (X3, Y3, Z3*Z).
+     *
+     * This is the effective affine technique: if we have a linear combination of group elements
+     * to compute, and all those group elements have the same Z coordinate, we can simply pretend
+     * that all those Z coordinates are 1, perform the computation that way, and then multiply the
+     * original Z coordinate back in.
+     *
+     * The technique works on any a=0 short Weierstrass curve. It is possible to generalize it to
+     * other curves too, but there the isomorphic curves will have different 'a' coefficients,
+     * which typically does affect the group laws.
+     *
+     *
+     * === Avoiding the square root for x-only point multiplication ===
+     *
+     * In this function, we want to compute the X coordinate of q*(n/d, y), for
+     * y = +-sqrt((n/d)^3 + 7).
+     *
+     * Let g = y^2*d^3 = n^3 + 7*d^3. This also means y = +-sqrt(g/d^3). Further note that
+     * sqrt(d*g) must exist if the input was valid, as d*g = y^2*d^4 = (y*d^2)^2.
+     *
+     * The input point (n/d, y) also has Jacobian coordinates:
+     *   (n/d, y, 1)
+     * = (n/d * sqrt^2(d*g), y * sqrt^3(d*g), sqrt(d*g))
+     * = (n/d * d*g, y * +-sqrt(d^3*g^3), sqrt(d*g))
+     * = (n/d * d*g, +-sqrt(g/d^3) * +-sqrt(d^3*g^3), sqrt(d*g))
+     * = (n*g, +-sqrt(g/d^3 * d^3*g^3), sqrt(d*g))
+     * = (n*g, +-sqrt(g^4), sqrt(d*g))
+     * = (n*g, +-g^2, sqrt(d*g))
```

Comment:
```suggestion
     * The input point (n/d, y) also has Jacobian coordinates:
     *   (n/d, y, 1)
     * = (n/d * sqrt^2(d*g), y * sqrt^3(d*g), sqrt(d*g))
     * = (n/d * d*g, y * sqrt(d^3*g^3), sqrt(d*g))
     * = (n/d * d*g, +-sqrt(g/d^3) * sqrt(d^3*g^3), sqrt(d*g))
     * = (n*g, +-sqrt(g/d^3 * d^3*g^3), sqrt(d*g))
     * = (n*g, +-sqrt(g^4), sqrt(d*g))
     * = (n*g, +-g^2, sqrt(d*g))
```

Reply:
Ok, yes... 

(indepedent nit: adding more space left to the equation would make it more readable. The `*` from the comments make it hard to read.)